Compile on one rank support: Redistribute targets

aorenste opened the pull request:
UPDATE: This is being replaced by the "collective targets as graph inputs" project. 

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172190
* #172161
* #173377
* __->__ #167553



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

aorenste committed:
WIP: DTensor Redistribute: impl

[ghstack-poisoned]

aorenste commented:
> naive question: why is the PR titled DTensor redisribute impl? if you could update the PR desc it would help for understanding motivation. the code doesn't appear all that related to dtensor, but it looks like its related to process group representation in traces?

It's not really ready for review yet. Before it's ready for review I'll fix the title/summary.

aorenste committed:
Update on "WIP: DTensor Redistribute: impl"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "WIP: DTensor Redistribute: impl"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "WIP: DTensor Redistribute: impl"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "WIP: DTensor Redistribute: impl"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "WIP: DTensor Redistribute: impl"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste commented on torch/distributed/device_mesh.py:
```diff
@@ -448,6 +448,17 @@ def _init_one_process_group(
+            # Couldn't we just tell c10d to use the same name on every rank? In
+            # theory yes, but for consistency we want to create ALL groups (even
+            # ones that don't contain our rank) and there's checks to ensure
+            # that we don't duplicate names.
```
Should this behavior be behind a config?

xmfan commented on torch/distributed/device_mesh.py:
```diff
@@ -456,16 +467,26 @@ def _init_one_process_group(
+                    if pg_name != dim_group.group_name:
+                        torch._C._distributed_c10d._register_process_group_alias(
+                            pg_name, dim_group.group_name
+                        )
```
Btw, you can also create process groups directly w/ torch.distributed.init_process_group. I don't think that codepath goes through DeviceMesh.

aorenste committed:
Update on "Compile on one rank support: Redistribute"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "Compile on one rank support: Redistribute"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "Compile on one rank support: Redistribute targets"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented:
In VC chats you mentioned that there were many other strategies you tried that did not work before you settled on this one. It would be very helpful to see them written down here.

aorenste committed:
Update on "Compile on one rank support: Redistribute targets"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "Compile on one rank support: Redistribute targets"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

aorenste committed:
Update on "Compile on one rank support: Redistribute targets"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Compile on one rank support: Redistribute targets"

DTensor.redistribute decomposes into collective operations which have a target string. This target string is different on every rank but refers to the same underlying C++ ProcessGroup.

This PR fixes this by:
- Adds an alias table for ProcessGroup (`_register_process_group_alias`)
- When registering ProcessGroups (in device_mesh.py) we register an alias from the "group name" to the individual ProcessGroup for that rank. The "group name" is just the ProcessGroup name for the first rank in the dimension.
- Instead of returning the current rank's ProcessGroup name we use the group name for the DeviceMesh ProcessGroup name.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
